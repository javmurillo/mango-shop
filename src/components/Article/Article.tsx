import { Badge, Card } from 'react-bootstrap';
import styled from 'styled-components';
import * as ArticleTypes from '../../store/types/articles.types';

const StyledPriceWrapper = styled.div`
  display: flex;
  justify-content: center;
`;

const StyledCurrentPrice = styled.span`
  color: green;
  font-size: 20px;
  margin-left: 6px;
`;

const StyledOldPrice = styled.span`
  color: darkred;
  text-decoration: line-through;
`;

export const Article = (props: { article: ArticleTypes.Article }) => {
  const { article } = props;
  return (
    <Card style={{ width: '16rem', margin: '2rem' }}>
      <Card.Img variant="top" src={article.img} />
      <Card.Body>
        <Card.Title>{article.name}</Card.Title>
        <Card.Text>{article.description}</Card.Text>
      </Card.Body>
      <Card.Footer
        className="text-muted"
        style={{
          padding: '2px',
          background: 'transparent',
          textAlign: 'center',
          border: 'none',
        }}
      >
        {article.oldPrice ? <Badge variant="success">¡Oferta!</Badge> : null}
      </Card.Footer>
      <Card.Footer className="text-muted">
        <StyledPriceWrapper>
          {article.oldPrice ? (
            <StyledOldPrice>{article.oldPrice} €</StyledOldPrice>
          ) : null}

          <StyledCurrentPrice>{article.currentPrice} €</StyledCurrentPrice>
        </StyledPriceWrapper>
      </Card.Footer>
    </Card>
  );
};
