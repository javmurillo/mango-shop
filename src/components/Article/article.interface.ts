export interface ArticleDto {
    name: string;
}