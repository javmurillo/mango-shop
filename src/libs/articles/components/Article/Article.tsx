import { Badge, Card } from 'react-bootstrap';
import styled from 'styled-components';
import { ArticleProps } from './article-props.interface';

const StyledPriceWrapper = styled.div`
  display: flex;
  justify-content: center;
`;

const StyledCurrentPrice = styled.span`
  color: green;
  font-size: 20px;
  margin-left: 6px;
`;

const StyledOldPrice = styled.span`
  color: darkred;
  text-decoration: line-through;
`;

const StyledCard = styled(Card)`
  width: 16rem;
  margin: 2rem;
`;

const StyledCardFooter = styled(Card.Footer)`
  padding: 2px;
  background: transparent;
  text-align: center;
  border: none;
`;

/**
 * Article component which displays the article information.
 * @param props ArticleDto
 */
export const Article = (props: ArticleProps): JSX.Element => {
  const { article } = props;
  return (
    <StyledCard aria-label="article-card">
      <Card.Img variant="top" src={article.img} />
      <Card.Body>
        <Card.Title>{article.name}</Card.Title>
        <Card.Text>{article.description}</Card.Text>
      </Card.Body>
      <StyledCardFooter>
        {article.oldPrice ? (
          <Badge variant="success">Deal of the day!</Badge>
        ) : null}
      </StyledCardFooter>
      <Card.Footer className="text-muted">
        <StyledPriceWrapper>
          {article.oldPrice ? (
            <StyledOldPrice>{article.oldPrice} €</StyledOldPrice>
          ) : null}
          <StyledCurrentPrice>{article.currentPrice} €</StyledCurrentPrice>
        </StyledPriceWrapper>
      </Card.Footer>
    </StyledCard>
  );
};
