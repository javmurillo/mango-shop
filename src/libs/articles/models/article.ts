export interface Article {
  id: string;
  name: string;
  description: string;
  img: string;
  currentPrice: number;
  oldPrice?: number;
}
