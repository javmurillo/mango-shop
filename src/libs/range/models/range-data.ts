export interface RangeData {
  min: number;
  max: number;
  rangeValues: number[];
}
