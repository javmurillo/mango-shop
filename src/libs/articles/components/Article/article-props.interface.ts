import { Article } from '../../models/article';

export interface ArticleProps {
  article: Article;
  ariaLabel?: string;
}
