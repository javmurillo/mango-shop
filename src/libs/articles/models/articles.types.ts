import { Action } from 'redux';
import { Article } from './article';

export const SET_ARTICLES = 'SET_ARTICLES';
export const FETCH_ARTICLES_FAILED = 'FETCH_ARTICLES_FAILED';
export const FILTER_ARTICLES = 'FILTER_ARTICLES';

export interface SetArticlesAction extends Action {
  type: typeof SET_ARTICLES;
  payload: { articles: Article[] };
}

export interface FetchArticlesFailedAction extends Action {
  type: typeof FETCH_ARTICLES_FAILED;
}

export interface FilterArticlesAction extends Action {
  type: typeof FILTER_ARTICLES;
  payload: { min: number; max: number };
}

export type ArticleAction =
  | SetArticlesAction
  | FetchArticlesFailedAction
  | FilterArticlesAction;

export type ArticleDispatchType = (args: ArticleAction) => ArticleAction;
