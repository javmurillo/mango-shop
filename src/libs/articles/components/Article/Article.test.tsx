import { cleanup, render } from '@testing-library/react';
import TestRenderer from 'react-test-renderer';
import { Article as ArticleDto } from '../../models/article';
import { Article } from './Article';

afterEach(cleanup);

const setup = () => {
  const utils = render(
    <Article article={testArticle} ariaLabel="article-card" />
  );
  const input = utils.getByLabelText('article-card');
  return {
    input,
    ...utils,
  };
};

describe('<Article/> tests', () => {
  test('renders correctly', () => {
    const { input } = setup();
    expect(input).toBeInTheDocument();
  });

  test('shoud match snapshot', () => {
    const tree = TestRenderer.create(
      <Article article={testArticle} />
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });
});

const testArticle: ArticleDto = {
  id: '1',
  name: 'Test article',
  currentPrice: 75,
  oldPrice: 90,
  description: 'Test description',
  img: 'Test image',
};
