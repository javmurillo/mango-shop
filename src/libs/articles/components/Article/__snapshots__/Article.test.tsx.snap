// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Article/> tests shoud match snapshot 1`] = `
<div
  aria-label="article-card"
  className="sc-hKgILt iRatjO card"
>
  <img
    className="card-img-top"
    src="Test image"
  />
  <div
    className="card-body"
  >
    <div
      className="card-title h5"
    >
      Test article
    </div>
    <p
      className="card-text"
    >
      Test description
    </p>
  </div>
  <div
    className="sc-eCssSg bUELJj card-footer"
  >
    <span
      className="badge badge-success"
    >
      Deal of the day!
    </span>
  </div>
  <div
    className="text-muted card-footer"
  >
    <div
      className="sc-bdfBwQ dHJFwu"
    >
      <span
        className="sc-dlfnbm hsBRpN"
      >
        90
         €
      </span>
      <span
        className="sc-gsTCUz dACZDE"
      >
        75
         €
      </span>
    </div>
  </div>
</div>
`;
